### ISSUE TITLE
How should people install 3rd party plugins outside of certbot-auto?

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#5731]
Currently the way Certbot plugins work is we use [setuptools entry_points](https://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins). This means the plugin must be installed at a path in [sys.path](https://docs.python.org/2/library/sys.html#sys.path).

For this reason, most of our 3rd party plugins tell people to run `(sudo) pip install` outside of a virtual environment to get their plugin. This isn't great as it can cause conflicts with packages installed from the users OS. Telling 3rd party plugins they need to get packaged by the distro also isn't a great option.

How can we reduce headaches here? Could Certbot have a canonical place for a virtual environment that Certbot also searches if its permissions are sane?

For context, the reason this is a bigger issue now is you need a DNS plugin for each provider so lots of new plugins are getting written and we don't want to take on the maintenance burden of all of them.

**Comment by Fmstrat** [comment]
I'm switching over to wildcards, and started doing this to resolve the conflict issue:

```
docker run -v /data/certs2/letsencrypt/:/etc/letsencrypt -v /root/ns1.ini:/ns1.ini:ro -it --name certbot --entrypoint "/bin/sh" -d certbot/certbot 
docker exec -it certbot pip install certbot-dns-nsone
docker exec -it certbot certbot certonly --agree-tos --non-interactive --email letsencrypt@ADOMAIN.com --dns-nsone --dns-nsone-credentials /ns1.ini --dns-nsone-propagation-seconds 60 --server https://acme-v02.api.letsencrypt.org/directory --debug -v -d *.MYDOMAIN.com -d MYDOMAIN.com
docker stop certbot
docker rm certbot
```

Of course, I'm about to report an issue with use of the NS1 API because it needs to change to support wildcards (I think).

**Comment by bmw** [comment]
@Fmstrat, we have prebuilt Docker images for the DNS plugins available if you want them. For example, here's the one for NS1: https://hub.docker.com/r/certbot/dns-nsone/

As for the issue, is it the one described by #5673? If not, please do open an issue.

**Comment by Fmstrat** [comment]
@bmw, man, that should really be added to the documentation! I never would have found that on docker's hub. Thanks for the tip. And yes, that pasting above was in reference to the issue I created which you linked to the Lexicon one. I was just including it to show the poster how to isolate from the host, however your docker image is a much better solution.

**Comment by bmw** [comment]
I believe we have added it to the documentation now at https://certbot.eff.org/docs/install.html#running-with-docker, but we definitely were scrambling to get all the pieces in place after the ACMEv2 launch and the documentation lagged behind by a day or so. If you found out about our Docker image elsewhere and our DNS plugins aren't mentioned, please let me know so I can try and get it updated.

**Comment by Fmstrat** [comment]
@bmw, the main documentation for the DNS plugins is here: https://certbot.eff.org/docs/using.html#dns-plugins

In the second paragraph, there is a link for "Docker" that points here: https://certbot.eff.org/docs/contributing.html#docker which does not have the same instructions as the link above.

Given that it moves from the user docs to contribution docs, I'd recommend both updating the contribution docs and pointing the first link I mentioned to https://certbot.eff.org/docs/install.html#running-with-docker since that's more focused on the end user than the contribution page is.

Thanks!

**Comment by erikrose** [comment]
~~pkg_resources has a class called `WorkingSet` which holds the `iter_entry_points()` method, among others. It is implicitly instantiated to provide the global entrypoint-resolving routines we're all familiar with. However, I suspect we could instantiate our own, pointing at a different venv (or even non-venv, if we could find a nice way of installing things into it), and let people drop plugins into that. The rest of the machinery in pkg_resources looks like it handles all the variants of module-loading you could ever want. Will this work at all? We should play with a proof of concept.~~

We could do this easily by just adding a new folder to sys.path, either in some conventional place or in one pointed to by a piece of configuration.

Here's the tricky part: whatever plugins we install in the venv will bring along dependencies. If we don't add the venv to sys.path, the dependencies won't be found. If we prepend it, they'll shadow the ones in whatever env or venv certbot is running from. That might be acceptable, except that one of those dependencies is bound to be certbot itself, and we certainly don't want an un-hash-checked version of certbot coming in and taking precedence. However, certbot itself will have already been loaded in sys.modules by the time the venv is in play, so it itself won't be shadowed, but there is the potential for any of certbot's dependencies that haven't been imported yet to be shadowed. I wouldn't want to debug that nondeterminism. Thus, I think we'd have to put the new venv on sys.path *after* certbot's env. Consequently, third-party plugins will not effectively be able to require versions of libs newer than those in certbot itself. That seems an acceptable tradeoff to me, and of course we should make it explicit in the docs. This also softens the blow of un-hash-checkedness: only the plugin and any new packages it depends on would be unchecked. (That's still pretty wide-open, of course, but I think all our security guarantees are pretty much moot once you start plugging in third-party code as root.)

There's [some precedent](https://github.com/jsvine/envplus) for this sort of thing, which suggests there are no serious gotchas.

And the same technique should work, regardless of whether certbot-auto is in play.

**Comment by maximillianh** [comment]
This is being addressed in #6613.
